From a89533c1e1be4ca9ed08baf7ee6be3fdcb8acb1e Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Mon, 17 Mar 2025 11:53:49 +0100
Subject: [PATCH] iio: trigger: rework to follow stm32mp25 upstream

Rework to follow changes done upstream on stm32mp25:
- simplify infrastructure by keeping devm_iio_trigger_register.
- Don't need to cast compatible data.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I0fa288930b79a6fe22104103afaca99c8d42e809
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/445395
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Olivier MOYSAN <olivier.moysan@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/iio/trigger/stm32-lptimer-trigger.c | 44 +++------------------
 1 file changed, 6 insertions(+), 38 deletions(-)

--- a/drivers/iio/trigger/stm32-lptimer-trigger.c
+++ b/drivers/iio/trigger/stm32-lptimer-trigger.c
@@ -53,7 +53,6 @@ static const struct stm32_lptim_cfg stm3
 struct stm32_lptim_trigger {
 	struct device *dev;
 	const char * const *triggers;
-	struct list_head tr_list;
 };
 
 static int stm32_lptim_validate_device(struct iio_trigger *trig,
@@ -82,21 +81,11 @@ bool is_stm32_lptim_trigger(struct iio_t
 }
 EXPORT_SYMBOL(is_stm32_lptim_trigger);
 
-static void stm32_lptim_unregister_triggers(struct stm32_lptim_trigger *priv)
-{
-	struct iio_trigger *tr;
-
-	list_for_each_entry(tr, &priv->tr_list, alloc_list)
-		iio_trigger_unregister(tr);
-}
-
-static int stm32_lptim_register_triggers(struct stm32_lptim_trigger *priv)
+static int stm32_lptim_setup_trig(struct stm32_lptim_trigger *priv)
 {
 	const char * const *cur = priv->triggers;
 	int ret;
 
-	INIT_LIST_HEAD(&priv->tr_list);
-
 	while (cur && *cur) {
 		struct iio_trigger *trig;
 
@@ -108,11 +97,9 @@ static int stm32_lptim_register_triggers
 		trig->ops = &stm32_lptim_trigger_ops;
 		iio_trigger_set_drvdata(trig, priv);
 
-		ret = iio_trigger_register(trig);
+		ret = devm_iio_trigger_register(priv->dev, trig);
 		if (ret)
 			return ret;
-
-		list_add_tail(&trig->alloc_list, &priv->tr_list);
 		cur++;
 	}
 
@@ -124,7 +111,6 @@ static int stm32_lptim_trigger_probe(str
 	struct stm32_lptim_trigger *priv;
 	struct stm32_lptim_cfg const *lptim_cfg;
 	u32 index;
-	int ret;
 
 	priv = devm_kzalloc(&pdev->dev, sizeof(*priv), GFP_KERNEL);
 	if (!priv)
@@ -141,37 +127,19 @@ static int stm32_lptim_trigger_probe(str
 	priv->dev = &pdev->dev;
 	priv->triggers = lptim_cfg->triggers[index];
 
-	ret = stm32_lptim_register_triggers(priv);
-	if (ret) {
-		stm32_lptim_unregister_triggers(priv);
-		return ret;
-	}
-
-	platform_set_drvdata(pdev, priv);
-
-	return 0;
-}
-
-static int stm32_lptim_trigger_remove(struct platform_device *pdev)
-{
-	struct stm32_lptim_trigger *priv = platform_get_drvdata(pdev);
-
-	stm32_lptim_unregister_triggers(priv);
-
-	return 0;
+	return stm32_lptim_setup_trig(priv);
 }
 
 static const struct of_device_id stm32_lptim_trig_of_match[] = {
-	{ .compatible = "st,stm32-lptimer-trigger", .data = (void *)&stm32mp15_lptim_cfg },
-	{ .compatible = "st,stm32mp21-lptimer-trigger", .data = (void *)&stm32mp25_lptim_cfg},
-	{ .compatible = "st,stm32mp25-lptimer-trigger", .data = (void *)&stm32mp25_lptim_cfg},
+	{ .compatible = "st,stm32-lptimer-trigger", .data = &stm32mp15_lptim_cfg },
+	{ .compatible = "st,stm32mp21-lptimer-trigger", .data = &stm32mp25_lptim_cfg},
+	{ .compatible = "st,stm32mp25-lptimer-trigger", .data = &stm32mp25_lptim_cfg},
 	{},
 };
 MODULE_DEVICE_TABLE(of, stm32_lptim_trig_of_match);
 
 static struct platform_driver stm32_lptim_trigger_driver = {
 	.probe = stm32_lptim_trigger_probe,
-	.remove = stm32_lptim_trigger_remove,
 	.driver = {
 		.name = "stm32-lptimer-trigger",
 		.of_match_table = stm32_lptim_trig_of_match,
